packages/components/src/tools-panel/tools-panel-item/hook.ts

```diff
@@ -129,8 +131,22 @@ export function useToolsPanelItem(
 			shouldRenderPlaceholder &&
 			! isShown &&
 			styles.ToolsPanelItemPlaceholder;
-		return cx( styles.ToolsPanelItem, placeholderStyle, className );
-	}, [ isShown, shouldRenderPlaceholder, className ] );
+		const firstItemStyle = firstDisplayedItem === label && 'first';
+		const lastItemStyle = lastDisplayedItem === label && 'last';
```

Comment:
> If the problem is only regarding the interaction with ItemGroup, maybe we could keep the solution in the @wordpress/components package?

I'm not sure that works but could be wrong. 

The issue I'm trying to tackle is the styling of the first and last **_visible_** elements. Normally, in an item group, all items are displayed, so pseudo-selectors can easily target the first and last items and style them appropriately. With the `ToolsPanel` we can have items rendering placeholders that aren't visible (this is to maintain the ordering of SlotFill injected items). These invisible items prevent the pseudo-selectors from being a solution.

Is it not possible that an ItemGroup could be styled externally to the `wordpress/components` package? In which case, to keep consistency of styling, the consumer needs to be able to target to the first and last visible elements of the ToolsPanel as well. Meaning, dynamic class names here wouldn't allow that.

Reply:
> Would it be acceptable to allow consumers of the ToolsPanel to provide these first and last class names via two new props?

This would definitely work, but I'd like to keep it as a fallback solution for a few reasons:

- adding new props increases the public API surface of the component, which means more code that we have to maintain and make sure doesn't break across releases
- having too many props can also hurt/overwhelm the developer experience
- it would be interesting to see if we can harness the components styling system and/or the context system

Could you prepare a little Storybook playground that illustrates the problem we're trying to solve? It would make it easier for me to collaborate on a solution. Thank you very much!

Reply:
I'm happy to put together something in a story for this. In the meantime, you can see the real-world use case in https://github.com/WordPress/gutenberg/pull/34027.

> but I'd like to keep it as a fallback solution for a few reasons:

Understood. This wouldn't have been my first pick either however it is all that I have left in terms of ideas that meets the "no hardcoded classes" criteria.

If `ItemGroup` is never styled differently externally to the components package, I'd say that we could add the first/last styling of visible `ToolsPanelItem`s within the `ToolsPanel` component. I don't believe that would be a fair assumption though.

If consumers need to style the first and last displayed items (which are not targetable via CSS pseudo-selectors) they need to know ahead of time selectors for those. I still see either the `ToolsPanel` needing to provide classes or the consumer being able to set them. I didn't think the context system had been adopted to the extent it could solve the issue but I am hopefully missing something there.

Reply:
I've added a [storybook example](https://github.com/WordPress/gutenberg/pull/37546/commits/3ded5006d8f0e62a531e6320910ae499e0061b47) attempting to replicate the use case from https://github.com/WordPress/gutenberg/pull/34027.

1. Run `npm run storybook:dev`
2. View [ToolsPanelWithItemGroup example](http://localhost:50240/?path=/story/components-experimental-toolspanel--tools-panel-with-item-group-slot)

<img src="https://user-images.githubusercontent.com/60436221/148728603-a4a7dfef-6ce0-4966-9d3d-6fd4f5030a00.gif" width="400"/>

Reply:
Thank you for adding an example, that makes it much easier to understand.

I think I've managed to achieve a similar effect while avoiding the `first` and `last` classnames

<details>

<summary>Click to expand</summary>

```diff
diff --git a/packages/components/src/item-group/item-group/component.tsx b/packages/components/src/item-group/item-group/component.tsx
index e58eae8976..0e05162e7b 100644
--- a/packages/components/src/item-group/item-group/component.tsx
+++ b/packages/components/src/item-group/item-group/component.tsx
@@ -10,7 +10,7 @@ import type { Ref } from 'react';
 import { contextConnect, WordPressComponentProps } from '../../ui/context';
 import { useItemGroup } from './hook';
 import { ItemGroupContext, useItemGroupContext } from '../context';
-import { View } from '../../view';
+import { ItemGroup as ItemGroupView } from '../styles';
 import type { ItemGroupProps } from '../types';
 
 function ItemGroup(
@@ -36,7 +36,7 @@ function ItemGroup(
 
 	return (
 		<ItemGroupContext.Provider value={ contextValue }>
-			<View { ...otherProps } ref={ forwardedRef } />
+			<ItemGroupView { ...otherProps } ref={ forwardedRef } />
 		</ItemGroupContext.Provider>
 	);
 }
diff --git a/packages/components/src/item-group/styles.ts b/packages/components/src/item-group/styles.ts
index 131754cfa5..2679384a34 100644
--- a/packages/components/src/item-group/styles.ts
+++ b/packages/components/src/item-group/styles.ts
@@ -2,11 +2,13 @@
  * External dependencies
  */
 import { css } from '@emotion/react';
+import styled from '@emotion/styled';
 
 /**
  * Internal dependencies
  */
 import { CONFIG, COLORS } from '../utils';
+import { View } from '../view';
 
 export const unstyledButton = css`
 	appearance: none;
@@ -27,6 +29,8 @@ export const unstyledButton = css`
 	}
 `;
 
+export const ItemGroup = styled( View )``;
+
 export const itemWrapper = css`
 	width: 100%;
 	display: block;
diff --git a/packages/components/src/tools-panel/stories/tools-panel-with-item-group-slot.js b/packages/components/src/tools-panel/stories/tools-panel-with-item-group-slot.js
index 78949c9818..9415e6a486 100644
--- a/packages/components/src/tools-panel/stories/tools-panel-with-item-group-slot.js
+++ b/packages/components/src/tools-panel/stories/tools-panel-with-item-group-slot.js
@@ -209,21 +209,6 @@ const SlotWrapper = css`
 `;
 
 const ToolsPanelItemClass = css`
-	border-left: 1px solid rgba( 0, 0, 0, 0.1 );
-	border-right: 1px solid rgba( 0, 0, 0, 0.1 );
-	border-bottom: 1px solid rgba( 0, 0, 0, 0.1 );
-
-	&.first {
-		border-top-left-radius: 20px;
-		border-top-right-radius: 20px;
-		border-top: 1px solid rgba( 0, 0, 0, 0.1 );
-	}
-
-	&.last {
-		border-bottom-left-radius: 20px;
-		border-bottom-right-radius: 20px;
-	}
-
 	&& > div,
 	&& > div > button {
 		border-radius: inherit;
diff --git a/packages/components/src/tools-panel/styles.ts b/packages/components/src/tools-panel/styles.ts
index 0819ca0c53..3286984924 100644
--- a/packages/components/src/tools-panel/styles.ts
+++ b/packages/components/src/tools-panel/styles.ts
@@ -13,6 +13,7 @@ import {
 	Wrapper as BaseControlWrapper,
 } from '../base-control/styles/base-control-styles';
 import { LabelWrapper } from '../input-control/styles/input-control-styles';
+import { ItemGroup } from '../item-group/styles';
 import { COLORS, CONFIG } from '../utils';
 import { space } from '../ui/utils/space';
 
@@ -147,6 +148,12 @@ export const ToolsPanelItem = css`
 	${ BaseControlLabel } {
 		line-height: 1.4em;
 	}
+
+	${ ItemGroup } & {
+		border-left: 1px solid rgba( 0, 0, 0, 0.1 );
+		border-right: 1px solid rgba( 0, 0, 0, 0.1 );
+		border-bottom: 1px solid rgba( 0, 0, 0, 0.1 );
+	}
 `;
 
 export const ToolsPanelItemPlaceholder = css`
@@ -156,3 +163,18 @@ export const ToolsPanelItemPlaceholder = css`
 export const DropdownMenu = css`
 	min-width: 200px;
 `;
+
+export const ToolsPanelItemFirst = css`
+	${ ItemGroup } & {
+		border-top-left-radius: 20px;
+		border-top-right-radius: 20px;
+		border-top: 1px solid rgba( 0, 0, 0, 0.1 );
+	}
+`;
+
+export const ToolsPanelItemLast = css`
+	${ ItemGroup } & {
+		border-bottom-left-radius: 20px;
+		border-bottom-right-radius: 20px;
+	}
+`;
diff --git a/packages/components/src/tools-panel/tools-panel-item/hook.ts b/packages/components/src/tools-panel/tools-panel-item/hook.ts
index 2e1299d0d8..4248cbb1d4 100644
--- a/packages/components/src/tools-panel/tools-panel-item/hook.ts
+++ b/packages/components/src/tools-panel/tools-panel-item/hook.ts
@@ -131,8 +131,10 @@ export function useToolsPanelItem(
 			shouldRenderPlaceholder &&
 			! isShown &&
 			styles.ToolsPanelItemPlaceholder;
-		const firstItemStyle = firstDisplayedItem === label && 'first';
-		const lastItemStyle = lastDisplayedItem === label && 'last';
+		const firstItemStyle =
+			firstDisplayedItem === label && styles.ToolsPanelItemFirst;
+		const lastItemStyle =
+			lastDisplayedItem === label && styles.ToolsPanelItemLast;
 		return cx(
 			styles.ToolsPanelItem,
 			placeholderStyle,
```
</details>

What do you think about this approach?

---

This storybook example raises another point — the fact that `Item` is not being used as a direct descendant/child of `ItemGroup`.

In fact, as of now `ItemGroup` assumes that `Item` is a direct descendant (see for example [these styles here](https://github.com/WordPress/gutenberg/blob/341473bb0949893aeb8c658cc255898745bb7425/packages/components/src/item-group/styles.ts#L45-L73)).

Using `Item` and `ItemGroup` like in this storybook example could potentially lead to unexpected results. For example, `ItemGroup` renders an element with role `list`, `Item` an element with role `listitem`. I'm not sure about if this semantically still works when they're not direct descendants.

Is there any way we can have `ItemGroup` and `Item`s as direct parent/children in the DOM structure (the `as` prop could be useful)?

Reply:
The driving force for all this was because the color panel was updated to use the `ItemGroup` [without considering the open PR ](https://github.com/WordPress/gutenberg/pull/34027)to change it to the ToolsPanel. The ToolsPanel is needed to control the display of color options especially as we are trying to expand color choices e.g. secondary/striped colors in tables, hover states for buttons etc.

Given the Color panel was updated as it was, https://github.com/WordPress/gutenberg/pull/34027 was updated to try and faithfully follow that decision. In doing so, I ran into problems using the `as` prop and `Item` for the `ToolsPanelItem`s. I will revisit that and see if I can make it work.

> What do you think about this approach? (actually an even better approach could be to add those styles to ItemGroup)

This doesn't address the [potential issue](https://github.com/WordPress/gutenberg/pull/37546#discussion_r780849809) I've stated [previously](https://github.com/WordPress/gutenberg/pull/37546#discussion_r779985192) about consumers styling item groups differently. As item groups use pseudo-selectors to target first and last items for styling, any consumer could just add their own styles to target them. Dynamic classes here will not allow consumers to target them easily. 

It has been stated previously we need to expect that components in the `wordpress/components` package can and will be used to build other UIs outside of WordPress.

Are we to expect consumers, in this case, to know to target the `ToolsPanelItemFirst`/`ToolsPanelItemLast` rules by partially matching class names via something like `[class*="ToolsPanelItemFirst"]`. Is this much different than a hardcoded class? To me, it might be worse, we still have to maintain those dynamic classes and it's more obscure, difficult to style etc.

Are we to ignore that consumers might wish to style the item groups differently, against the goal that components can be used for any UI? I'm obviously missing something but it feels a bit like here we are letting our ideology block our progress.

Also, I appreciate our desire to not introduce hardcoded classes given they become part of the API and must be maintained. I do not see the nature of these classes changing at all. The concepts of first or last visible items will always be the same. Any possible maintenance burden would be negligible.

Thanks for all your time looking at this and discussing the best path forward 👍

Reply:
> I will revisit that and see if I can make it work.

I've updated https://github.com/WordPress/gutenberg/pull/34027 as well as the storybook example (b02c4ef) in this PR to render the `ToolsPanelItem` as an `Item` component. I've also introduced a `Button` wrapper within the `Dropdown` toggle.

Now the `Item` is a direct child of the `ItemGroup` we should be free to decide on where to draw the line regarding the main question for this PR, that being whether consumers need to be able to style the first and last visible items in a `ToolsPanel`.

Reply:
> The ToolsPanel is needed to control the display of color options especially as we are trying to expand color choices

That is now clear, thank you for the extra context

> Are we to expect consumers, in this case, to know to target the ToolsPanelItemFirst/ToolsPanelItemLast rules by partially matching class names via something like [class*="ToolsPanelItemFirst"]. 

That wouldn't be a possibility, since what you see in the source code as, e.g. `ToolsPanelItemFirst`, in production at runtime gets hashed into a shorter classname (thus not exposing a hardcoded classname).

> Is this much different than a hardcoded class? To me, it might be worse, we still have to maintain those dynamic classes and it's more obscure, difficult to style etc.

The point that you raise here about maintaining it (similarly to a hardcoded classname) is a valid one, though.

> I ran into problems using the as prop and Item for the ToolsPanelItems. I will revisit that and see if I can make it work.
> ...
> Now the Item is a direct child of the ItemGroup we should be free to decide on where to draw the line regarding the main question for this PR

That's great news, this definitely unblocks this PR's main question.

> Are we to ignore that consumers might wish to style the item groups differently, against the goal that components can be used for any UI? I'm obviously missing something but it feels a bit like here we are letting our ideology block our progress.

We certainly don't want our ideology to block our progress, and I'm sorry if you've felt frustrated by my review comments 🙇 

Working on these components (which are not trivial on their own) and composing them in complex patterns is not an easy task. We are still in the process of converting legacy components to CSS in JS, adding Context support and finding out the best solution to overcome those composition challenges. While doing so, we need to be careful when introducing anything that external contributors can rely on (for example, a hardcoded classname), as that makes it harder to update components without introducing breaking changes.

In the light of the above, I would agree that **for now the best way forward would be to introduce two new props** (as you previously suggested), with the only caveat that they'd be marked as **experimental** for now (i.e. they are prefixed with `__experimental`). This will unblock this PR and any further work depending on it, while giving us the time to think about any alternative approach (if we find any).

Reply:
Thanks for the detailed reply @ciampo 👍 

> That wouldn't be a possibility, since what you see in the source code as, e.g. ToolsPanelItemFirst, in production at runtime gets hashed into a shorter classname (thus not exposing a hardcoded classname).

Good point.

> We certainly don't want our ideology to block our progress, and I'm sorry if you've felt frustrated by my review comments

Your reviews are always appreciated. 

My concerns here are with how strictly or absolutely we need to adhere to the no hard-coded class names rule. In this case, we've already spent much more time on this than any hypothetical maintenance burden in the future on something that won't change. 

Given the concept of a first and last visible item won't change, I'd still argue this would be an acceptable part of the API for this component. It could also be argued that the `Notice` component sets a similar precedent with the use of `is-${ status }` in its classes. These statuses could even be more likely to change than something more conceptually rigid like a first/last visible item. 

>  While doing so, we need to be careful when introducing anything that external contributors can rely on

I still appreciate the need for this and have happily made these changes previously. In this case though, these classes are something that external contributor might in fact need to rely on (or provide as previously suggested).

> as that makes it harder to update components without introducing breaking changes

Understood. As these classes wouldn't be changing nor the items they apply to, I'm not convinced they would make the component more difficult to update. I appreciate the caution and applied it myself before settling on the original approach here or offering the prop option.

> In the light of the above, I would agree that for now the best way forward would be to introduce two new props (as you previously suggested), with the only caveat that they'd be marked as experimental for now 

I've introduced the two new experimental props to allow for these classes to be supplied from the consumer.

> This will unblock this PR and any further work depending on it, while giving us the time to think about any alternative approach (if we find any).

Thank you for all the time, input and feedback on getting this unblocked. Appreciate it 🙇